File: services/btn-ml/ml_predictor.py
# Python script to serve ML predictions
# This script loads your trained model and provides a simple API

import pickle
import json
import sys
from typing import Dict, Any, Tuple

class WorkoutWeightPredictor:
    def __init__(self, model_path: str):
        """Load the trained model from pickle file"""
        try:
            with open(model_path, 'rb') as f:
                self.model_data = pickle.load(f)
            print(f"✓ Model loaded from {model_path}")
        except Exception as e:
            print(f"❌ Error loading model: {e}")
            self.model_data = None
    
    def predict_weights(self, features: Dict[str, Any]) -> Tuple[float, float]:
        """Predict male and female weights based on workout features"""
        if not self.model_data:
            # Fallback to heuristic if model not loaded
            return self._fallback_prediction(features)
        
        try:
            # Extract features in the same order as training
            feature_vector = self._extract_feature_vector(features)
            
            # Make predictions (assuming your model has separate male/female models)
            male_pred = self.model_data['male_model'].predict([feature_vector])[0]
            female_pred = self.model_data['female_model'].predict([feature_vector])[0]
            
            return float(male_pred), float(female_pred)
        except Exception as e:
            print(f"❌ Prediction error: {e}")
            return self._fallback_prediction(features)
    
    def _extract_feature_vector(self, features: Dict[str, Any]) -> list:
        """Convert features to the format expected by the model"""
        # Map movement types to numbers
        movement_map = {
            'olympic': 0, 'squat': 1, 'strength': 2, 
            'dumbbell': 3, 'pressing': 4, 'other': 5
        }
        
        # Map competition levels to numbers
        comp_map = {'Open': 0, 'Quarterfinals': 1, 'Semifinals': 2}
        
        # Map time domains to numbers
        time_map = {
            '1:00 - 5:00': 0, '5:00 - 10:00': 1, '10:00 - 15:00': 2,
            '15:00 - 20:00': 3, '20:00+': 4
        }
        
        # Map formats to numbers
        format_map = {'For Time': 0, 'AMRAP': 1, 'Rounds For Time': 2, 'Ladder': 3}
        
        return [
            movement_map.get(features.get('movement_type', 'other'), 5),
            features.get('total_reps', 0),
            features.get('largest_single_set', 0),
            features.get('time_cap_seconds', 600),
            time_map.get(features.get('time_domain', '5:00 - 10:00'), 1),
            format_map.get(features.get('format', 'For Time'), 0),
            features.get('total_weighted_movements', 1),
            features.get('has_other_barbell', 0),
            comp_map.get(features.get('competition_level', 'Open'), 0)
        ]
    
    def _fallback_prediction(self, features: Dict[str, Any]) -> Tuple[float, float]:
        """Fallback prediction using simple heuristics"""
        movement_type = features.get('movement_type', 'other')
        total_reps = features.get('total_reps', 25)
        
        # Base weights by movement type
        base_weights = {
            'olympic': (135, 95),
            'squat': (115, 75),
            'strength': (185, 125),
            'dumbbell': (50, 35),
            'pressing': (95, 65),
            'other': (100, 70)
        }
        
        male_base, female_base = base_weights.get(movement_type, (100, 70))
        
        # Adjust for rep count
        if total_reps > 100:
            multiplier = 0.7
        elif total_reps > 50:
            multiplier = 0.8
        elif total_reps > 25:
            multiplier = 0.9
        else:
            multiplier = 1.0
        
        return male_base * multiplier, female_base * multiplier

def main():
    """Simple command-line interface for testing"""
    if len(sys.argv) < 2:
        print("Usage: python ml_predictor.py <model_path>")
        sys.exit(1)
    
    model_path = sys.argv[1]
    predictor = WorkoutWeightPredictor(model_path)
    
    # Example usage
    example_features = {
        'movement_type': 'olympic',
        'total_reps': 45,
        'largest_single_set': 21,
        'time_cap_seconds': 600,
        'time_domain': '5:00 - 10:00',
        'format': 'For Time',
        'total_weighted_movements': 1,
        'has_other_barbell': 0,
        'competition_level': 'Open'
    }
    
    male_weight, female_weight = predictor.predict_weights(example_features)
    print(f"Predicted weights: {male_weight:.0f}/{female_weight:.0f} lbs")

if __name__ == "__main__":
    main()

File: lib/metcon_generator/mlPredictor.ts
// ML Weight Predictor Integration
// This module handles loading and using the trained ML model for weight prediction

export interface WorkoutFeatures {
  movement_type: string;
  total_reps: number;
  largest_single_set: number;
  time_cap_seconds: number;
  time_domain: string;
  format: string;
  total_weighted_movements: number;
  has_other_barbell: number;
  competition_level: string;
}

export interface WeightPrediction {
  male_weight: number;
  female_weight: number;
}

export interface MLResponse {
  male_weight: number;
  female_weight: number;
  model_used: boolean;
  features: WorkoutFeatures;
}

class MLWeightPredictor {
  private apiBaseUrl: string;
  private isHealthy: boolean = false;

  constructor(apiBaseUrl: string = 'http://localhost:5001') {
    this.apiBaseUrl = apiBaseUrl;
  }

  async checkHealth(): Promise<boolean> {
    try {
      const response = await fetch(`${this.apiBaseUrl}/health`);
      const data = await response.json();
      this.isHealthy = data.status === 'healthy' && data.model_loaded;
      return this.isHealthy;
    } catch (error) {
      console.warn('ML API health check failed:', error);
      this.isHealthy = false;
      return false;
    }
  }

  // Extract features from a generated workout
  extractFeatures(workout: any): WorkoutFeatures {
    const exercises = workout.exercises || [];
    
    // Determine movement type based on exercises
    const movementType = this.determineMovementType(exercises);
    
    // Calculate total reps
    const totalReps = exercises.reduce((sum: number, ex: any) => {
      if (typeof ex.reps === 'string') {
        // Handle rep schemes like "21-15-9"
        const repNumbers = ex.reps.split('-').map((r: string) => parseInt(r.trim()));
        return sum + repNumbers.reduce((a: number, b: number) => a + b, 0);
      }
      return sum + (ex.reps || 0);
    }, 0);

    // Find largest single set
    const largestSingleSet = exercises.reduce((max: number, ex: any) => {
      if (typeof ex.reps === 'string') {
        const repNumbers = ex.reps.split('-').map((r: string) => parseInt(r.trim()));
        return Math.max(max, ...repNumbers);
      }
      return Math.max(max, ex.reps || 0);
    }, 0);

    // Count weighted movements
    const weightedMovements = exercises.filter((ex: any) => 
      ex.weight || ex.weight_male_lbs || ex.weight_female_lbs
    ).length;

    // Check for other barbell exercises
    const hasOtherBarbell = exercises.some((ex: any) => 
      this.isBarbellExercise(ex.name) && ex.name !== exercises[0]?.name
    ) ? 1 : 0;

    return {
      movement_type: movementType,
      total_reps: totalReps,
      largest_single_set: largestSingleSet,
      time_cap_seconds: workout.time_cap_seconds || 600,
      time_domain: workout.time_domain || '5:00 - 10:00',
      format: workout.format || 'For Time',
      total_weighted_movements: weightedMovements,
      has_other_barbell: hasOtherBarbell,
      competition_level: 'Open' // Default to Open level
    };
  }

  private determineMovementType(exercises: any[]): string {
    const exerciseNames = exercises.map(ex => ex.name.toLowerCase());
    
    // Olympic lifts
    if (exerciseNames.some(name => 
      name.includes('snatch') || name.includes('clean') || name.includes('jerk')
    )) {
      return 'olympic';
    }
    
    // Squat movements
    if (exerciseNames.some(name => 
      name.includes('squat') || name.includes('thruster') || name.includes('overhead squat')
    )) {
      return 'squat';
    }
    
    // Strength movements
    if (exerciseNames.some(name => 
      name.includes('deadlift') || name.includes('bench press')
    )) {
      return 'strength';
    }
    
    // Dumbbell movements
    if (exerciseNames.some(name => 
      name.includes('dumbbell')
    )) {
      return 'dumbbell';
    }
    
    // Pressing movements
    if (exerciseNames.some(name => 
      name.includes('press') || name.includes('push')
    )) {
      return 'pressing';
    }
    
    return 'other';
  }

  private isBarbellExercise(exerciseName: string): boolean {
    const barbellExercises = [
      'Snatch', 'Deadlifts', 'Overhead Squats', 'Thrusters', 'Power Cleans', 
      'Clean and Jerks', 'Squat Cleans', 'Squat Snatch', 'Power Snatch'
    ];
    return barbellExercises.includes(exerciseName);
  }

  // Predict weights using ML API
  async predictWeights(features: WorkoutFeatures): Promise<WeightPrediction> {
    try {
      // Check API health first
      if (!this.isHealthy) {
        await this.checkHealth();
      }

      if (!this.isHealthy) {
        console.warn('ML API not available, using fallback prediction');
        return this.fallbackPrediction(features);
      }

      // Call ML API
      const response = await fetch(`${this.apiBaseUrl}/predict`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(features)
      });

      if (!response.ok) {
        throw new Error(`API request failed: ${response.status}`);
      }

      const data: MLResponse = await response.json();
      
      return {
        male_weight: data.male_weight,
        female_weight: data.female_weight
      };

    } catch (error) {
      console.warn('ML prediction failed, using fallback:', error);
      return this.fallbackPrediction(features);
    }
  }

  // Batch prediction for multiple workouts (useful for training blocks)
  async predictBatch(featuresArray: WorkoutFeatures[]): Promise<WeightPrediction[]> {
    try {
      if (!this.isHealthy) {
        await this.checkHealth();
      }

      if (!this.isHealthy) {
        console.warn('ML API not available, using fallback predictions');
        return featuresArray.map(features => this.fallbackPrediction(features));
      }

      const response = await fetch(`${this.apiBaseUrl}/predict/batch`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ workouts: featuresArray })
      });

      if (!response.ok) {
        throw new Error(`Batch API request failed: ${response.status}`);
      }

      const data = await response.json();
      
      return data.results.map((result: any) => ({
        male_weight: result.male_weight,
        female_weight: result.female_weight
      }));

    } catch (error) {
      console.warn('ML batch prediction failed, using fallback:', error);
      return featuresArray.map(features => this.fallbackPrediction(features));
    }
  }

  // Fallback prediction using heuristics
  private fallbackPrediction(features: WorkoutFeatures): WeightPrediction {
    const baseWeights = this.getBaseWeights(features.movement_type);
    const repAdjustment = this.calculateRepAdjustment(features.total_reps, features.largest_single_set);
    const competitionMultiplier = this.getCompetitionMultiplier(features.competition_level);
    
    const maleWeight = Math.round(baseWeights.male * repAdjustment * competitionMultiplier);
    const femaleWeight = Math.round(baseWeights.female * repAdjustment * competitionMultiplier);

    return {
      male_weight: maleWeight,
      female_weight: femaleWeight
    };
  }

  private getBaseWeights(movementType: string): { male: number; female: number } {
    const weights = {
      olympic: { male: 135, female: 95 },
      squat: { male: 115, female: 75 },
      strength: { male: 185, female: 125 },
      dumbbell: { male: 50, female: 35 },
      pressing: { male: 95, female: 65 },
      other: { male: 100, female: 70 }
    };
    return weights[movementType as keyof typeof weights] || weights.other;
  }

  private calculateRepAdjustment(totalReps: number, largestSet: number): number {
    // Higher reps = lighter weights
    if (totalReps > 100) return 0.7;
    if (totalReps > 50) return 0.8;
    if (totalReps > 25) return 0.9;
    if (totalReps > 10) return 1.0;
    return 1.1; // Low reps = heavier weights
  }

  private getCompetitionMultiplier(competitionLevel: string): number {
    const multipliers = {
      'Open': 1.0,
      'Quarterfinals': 1.1,
      'Semifinals': 1.2
    };
    return multipliers[competitionLevel as keyof typeof multipliers] || 1.0;
  }
}

export const mlPredictor = new MLWeightPredictor();